### PR TITLE
Merge wireprotocol and auth tests and wireprotocol fixes from PR#25 branch

### PR DESCRIPTION
**Pull request by jn9e9**
Plus fix to operations client ListAuthenticators and status code names and misc improvements to protobuf level client

Signed-off-by: John 9e9 <jn9e9e9@gmail.com>

### CODE REVIEW

File: interface/requests/response.go [hunk 1]
```diff
@@ -6,172 +6,187 @@ package requests
...
-	RespPsaErrorHardwareFailure      uint16 = 1147
-	RespPsaErrorInsufficientEntropy  uint16 = 1148
-	RespPsaErrorInvalidSignature     uint16 = 1149
-	RespPsaErrorInvalidPadding       uint16 = 1150
-	RespPsaErrorTamperingDetected    uint16 = 1151
+	StatusSuccess                      StatusCode = 0
+	StatusWrongProviderID              StatusCode = 1
+	StatusContentTypeNotSupported      StatusCode = 2
+	StatusAcceptTypeNotSupported       StatusCode = 3
+	StatusVersionTooBig                StatusCode = 4
```
Comment (hug-dev) [inline interface/requests/response.go:25]: For this one the spec has: `WireProtocolVersionNotSupported` ([here](https://parallaxsecond.github.io/parsec-book/parsec_client/status_codes.html))

File: interface/requests/response.go [hunk 2]
```diff
@@ -6,172 +6,187 @@ package requests
...
+	StatusProviderNotRegistered        StatusCode = 5
+	StatusProviderDoesNotExist         StatusCode = 6
+	StatusDeserializingBodyFailed      StatusCode = 7
+	StatusSerializingBodyFailed        StatusCode = 8
+	StatusOpcodeDoesNotExist           StatusCode = 9
+	StatusStatusonseTooLarge           StatusCode = 10
```
Comment (hug-dev) [inline interface/requests/response.go:31]: "Status Statusonse", sounds like a Roman emperor, great leader of the Parsec army 😄 
Maybe `StatusResponseTooLarge`?

File: interface/requests/response.go [hunk 3]
```diff
@@ -6,172 +6,187 @@ package requests
...
+	StatusUnsupportedOperation         StatusCode = 11
```
Comment (hug-dev) [inline interface/requests/response.go:32]: I think we deleted that one and removed 1 to the rest of the status code

File: interface/requests/response.go [hunk 4]
```diff
@@ -6,172 +6,187 @@ package requests
...
+	StatusAuthenticationError          StatusCode = 12
+	StatusAuthenticatorDoesNotExist    StatusCode = 13
+	StatusAuthenticatorNotRegistered   StatusCode = 14
+	StatusKeyDoesNotExist              StatusCode = 15
+	StatusKeyAlreadyExists             StatusCode = 16
```
Comment (hug-dev) [inline interface/requests/response.go:37]: Deleted those ones and added a few more! Check [here](https://parallaxsecond.github.io/parsec-book/parsec_client/status_codes.html)

File: interface/requests/response.go [hunk 5]
```diff
@@ -6,172 +6,187 @@ package requests
...
+	StatusPsaErrorInsufficientMemory   StatusCode = 1141
+	StatusPsaErrorInsufficientStorage  StatusCode = 1142
+	StatusPsaErrorInssuficientData     StatusCode = 1143
+	StatusPsaErrorCommunicationFailure StatusCode = 1145
+	StatusPsaErrorStorageFailure       StatusCode = 1146
+	StatusPsaErrorHardwareFailure      StatusCode = 1147
+	StatusPsaErrorInsufficientEntropy  StatusCode = 1148
+	StatusPsaErrorInvalidSignature     StatusCode = 1149
+	StatusPsaErrorInvalidPadding       StatusCode = 1150
+	StatusPsaErrorTamperingDetected    StatusCode = 1151
```
Comment (hug-dev) [inline interface/requests/response.go:56]: You might want to double check the PSA ones as well, we changed them from what they were at the beginning!

File: interface/requests/response.go [hunk 6]
```diff
@@ -6,172 +6,187 @@ package requests
...
+// StatusonseBody represents a Statusonse body
 type ResponseBody struct {
 	*bytes.Buffer
 }
 
 // Response represents a Parsec response
 type Response struct {
-	Header WireHeader
-	Body   ResponseBody
+	Header *wireHeader
```
Comment (hug-dev) [inline interface/requests/response.go:70]: Why the `*` here?

File: interface/requests/wireheader.go [hunk 7]
```diff
@@ -6,68 +6,165 @@ package requests
...
-	AcceptType   uint8
-	AuthType     uint8
-	BodyLen      uint32
-	AuthLen      uint16
-	OpCode       uint32
-	Status       uint16
-	Reserved1    uint8
-	Reserved2    uint8
+	versionMajor versionMajorType
+	versionMinor versionMinorType
```
Comment (hug-dev) [inline interface/requests/wireheader.go:52]: Note that the first four fields:
- magic number
- header size
- major version number
- minor version number

will always have the same width and position on every Parsec request/response header, no matter the wire protocol version. It would then make sense to either separate them in their own structure or not include them in the `wireHeader` structure, which will be the one that is versioned. In the [Rust interface](https://github.com/parallaxsecond/parsec-interface-rs/blob/master/src/requests/common/wire_header_1_0.rs), we put the rest in its own `wire_header_1_0` module to make it easy to write the logic selecting the correct wire header format. Though we did not implement that logic yet, at least we have the building blocks 😋 you have some more info [here](https://github.com/parallaxsecond/parsec/issues/169)

File: interface/requests/wireheader.go [hunk 8]
```diff
@@ -6,68 +6,165 @@ package requests
...
-	return nil
+	if r.hdrSize != wireHeaderSizeValue {
+		return nil, fmt.Errorf("invalid header size (%d != %d)", r.hdrSize, wireHeaderSizeValue)
+	}
+	r.versionMajor = versionMajorType(buf.Next(buffBytes8Bit)[0])
+	r.versionMinor = versionMinorType(buf.Next(buffBytes8Bit)[0])
+	if !isSupportedWireHeaderVersion(r.versionMajor, r.versionMinor) {
+		return nil, fmt.Errorf("unsupported version number %v.%v", r.versionMajor, r.versionMinor)
+	}
+	r.flags = flagsType(binary.LittleEndian.Uint16(buf.Next(buffBytes16Bit)))
```
Comment (hug-dev) [inline interface/requests/wireheader.go:135]: Imagine there were multiple wire protocol versions, then from line 134, you would need to use the correct parsing mechanism of the rest of the header depending on the version. Also the check line 127 might be worth to do here instead, the size of the header might change depending on the version.
I don't want to make your logic too complicated because of this, or too implement everything handling this now, as the wire protocol version might stay the same for a long time, but it's to make the little changes now that would make it easier to handle later

**Review by hug-dev — COMMENTED** [review]
Thanks for the enhancements and the tests ⭐ I put some general comments, which might not all be related to the actual changes 😬 If you don't want to do all in one go, we can still create issues for those and address them in future PRs.

File: interface/auth/auth_internal_test.go [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+		Context("Creating AuthDirect authenticator", func() {
+			BeforeEach(func() {
+				authType = AuthDirect
+			})
+			It("Should be a valid auth type", func() {
+				Expect(authType.IsValid()).To(BeTrue())
+			})
+			It("Should fail", func() {
+				Expect(err).To(HaveOccurred())
+				Expect(authenticator).To(BeNil())
```
[the remark this answers is not in the document]
Comment (jn9e9) [reply interface/auth/auth_internal_test.go:101]: currently not implemented, so testing that we don't fool a user - implementation ticket here: https://github.com/parallaxsecond/parsec-client-go/issues/8

File: interface/operations/operations_client.go [hunk 10]
```diff
@@ -82,6 +82,17 @@ func InitClient() (*Client, error) {
 	return client, nil
 }
 
+// InitClient initializes a Parsec client
+func InitClientFromConnection(conn connection.Connection) (*Client, error) {
+	client := &Client{
+		conn:     conn,
+		provider: requests.ProviderCore,
+		authType: auth.AuthUnixPeerCredentials,
```
[the remark this answers is not in the document]
Comment (jn9e9) [reply interface/operations/operations_client.go:90]: that sounds reasonable, but will fix under https://github.com/parallaxsecond/parsec-client-go/issues/6 i think - will reference this in the ticket though

File: interface/operations/operations_client.go [hunk 11]
```diff
@@ -533,10 +541,10 @@ func (c Client) operation(op requests.OpCode, provider requests.ProviderID, requ
 		return err
 	}
 
-	respBody, err := requests.NewResponse(rcvBuf, response)
+	respBody, err := requests.NewResponse(op, rcvBuf, response)
```
[the remark this answers is not in the document]
Comment (jn9e9) [reply interface/operations/operations_client.go:544]: yes, line 93 of response.go

File: interface/operations/operations_client.go [hunk 12]
```diff
@@ -155,7 +161,7 @@ func (c Client) ListKeys() ([]*listkeys.KeyInfo, error) {
 func (c Client) ListAuthenticators() ([]*listauthenticators.AuthenticatorInfo, error) {
 	req := &listauthenticators.Operation{}
 	resp := &listauthenticators.Result{}
-	err := c.operation(requests.OpListKeys, requests.ProviderCore, req, resp)
+	err := c.operation(requests.OpListAuthenticators, requests.ProviderCore, req, resp)
```
[the remark this answers is not in the document]
Comment (jn9e9) [reply interface/operations/operations_client.go:164]: assume you mean not take a provider?  Need the opcode to be different per operation.

File: interface/operations/operations_client.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (jn9e9) [reply interface/operations/operations_client.go:544]: the request and response are autgenerated protobuf code.  could do it by putting source files into the packages with the auto generated files, but would make clean prior to regeneration more of a fuss.  doable - will create an issue.

[the remark this answers is not in the document]
Comment (jn9e9) [reply interface/operations/operations_client.go:544]: i'm also not convinced it would reduce chance of mistakes - the pairing only happens in one place at the moment and we're not using the protocol buffer types elsewhere - we'd just move the 'get the wrong opcode' problem from a single file to a whole bunch of files

File: interface/requests/wireheader.go (hunk 7 above)
Comment (jn9e9) [reply interface/requests/wireheader.go:52]: created this issue: https://github.com/parallaxsecond/parsec-client-go/issues/31

File: interface/requests/wireheader.go (hunk 8 above)
Comment (jn9e9) [reply interface/requests/wireheader.go:135]: fair point, did thing about that at the time - have added to https://github.com/parallaxsecond/parsec-client-go/issues/31

File: interface/requests/response.go (hunk 6 above)
Comment (jn9e9) [reply interface/requests/response.go:70]: being a pointer makes the assignment on line 89 simpler as the parse method will return nil value for wireHeader if it fails - no other reason than slightly cleaner code and avoiding a copy (i think)

File: interface/operations/operations_client.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (hug-dev) [reply interface/operations/operations_client.go:164]: ah yes sorry, "not take a provider ID" is what I meant

File: interface/operations/operations_client.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (hug-dev) [reply interface/operations/operations_client.go:544]: I am not so good in Go, how does it work? The `request` and `response` fields seem to be of type `proto.Message` but you pass each specific generated `Operation` structure. Is it a cast?

I think it's better to leave it this way rather than modifying the generated code! It's not so much effort

File: interface/requests/wireheader.go [hunk 13]
```diff
@@ -6,68 +6,165 @@ package requests
...
 const (
 	buffBytes8Bit  int = 1
 	buffBytes16Bit int = 2
 	buffBytes32Bit int = 4
 	buffBytes64Bit int = 8
 )
 
-func (r *WireHeader) parse(buf *bytes.Buffer) error {
+func parseWireHeaderFromBuf(buf *bytes.Buffer) (*wireHeader, error) {
+	r := &wireHeader{}
```
Comment (hug-dev) [inline interface/requests/wireheader.go:117]: How does the memory management work here? To me it looks like you are creating a temporary empty `wireHeader` on the stack and assign `r` to its address. Then fill it with the values parsed and return `r`. But when leaving the function, isn't `r` essentially going to be a dangling pointer?
And similarly in line 82 of `response.go`: `r := &Response{}`

Comment (jn9e9) [reply interface/requests/wireheader.go:117]: this is worth a read: https://www.ardanlabs.com/blog/2017/05/language-mechanics-on-escape-analysis.html

TL;DR - the compiler detects if a pointer will escape the current function - if it will, it allocate the data structure on the heap, otherwise it is free to choose whether to allocate on stack or heap. (likely stack but no guarantees

File: interface/requests/response.go (hunk 6 above)
Comment (hug-dev) [reply interface/requests/response.go:70]: Ah I see. Well I was also going to propose to maybe change `parseWireHeaderFromBuf`
 to return a `wireHeader` instead? Then here we can have a `wireHeader` too to match the structure of the `Request`:
  ```go
type Request struct {
	Header wireHeader
	Body   RequestBody
	Auth   auth.RequestAuthToken
}
```
Or it's definitely more efficient to keep pointers everywhere?

Also, it is possible to keep the `Body` field? A `Response` is really a header + a body. If you think that is not needed here, I propose that we remove the `Response` struct altogether but I believe it would be nice to group them though.

Comment (hug-dev) [reply interface/requests/response.go:70]: I don't want to sound too directive!! If you think it's more idiomatic to have pointers, let's keep them there!!

The only thing important here I think is to group the body (into the appropriate type that you think) and the header into a `Response`

File: interface/requests/response.go (hunk 4 above)
Comment (jn9e9) [reply interface/requests/response.go]: fixed

File: interface/requests/response.go (hunk 1 above)
Comment (jn9e9) [reply interface/requests/response.go:25]: updated

File: interface/requests/response.go (hunk 2 above)
Comment (jn9e9) [reply interface/requests/response.go]: :) - fixed, find/replace error

File: interface/requests/response.go (hunk 3 above)
Comment (jn9e9) [reply interface/requests/response.go]: done

File: interface/requests/response.go (hunk 5 above)
Comment (jn9e9) [reply interface/requests/response.go]: re-created whole lot from docs

File: interface/requests/response.go (hunk 6 above)
Comment (jn9e9) [reply interface/requests/response.go]: i've removed the response and responsebody types and refactored the NewResponse into ParseResponse - none of this was used downstream and wasn't used.  No point in having code with no real purpose (and don't think it future proofs anything)

File: interface/operations/operations_client.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (jn9e9) [reply interface/operations/operations_client.go]: its how the go protobuf stuff works - basically passing an interface around - all the generated protobuf types implement the interface - it then uses reflection to do the (un)marshalling.

So the types I pass in are generated types implementing the interface spec.

File: interface/operations/operations_client.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (jn9e9) [reply interface/operations/operations_client.go]: Good idea, but when got into detail, was going to be quite messy to do it with the existing provider field of the operations.Client.

ended up using a strategy pattern using the operations.ProviderSelector interface to allow the operations.Client to call back to the caller (parsec.BasicClient) to get the best provider to use.  Also moved the provider crypto checks to parsec.BasicClient

[the remark this answers is not in the document]
Comment (hug-dev) [reply interface/operations/operations_client.go]: Arg, what kind of details? Sorry if I was confusing. I was thinking of something dead simple:
Keeping the `provider` and `authType` fields on `Client` and modifying `InitClient` to:
```go
func InitClient(provider requests.ProviderID, authType auth.AuthenticationType) (*Client, error) {
	conn, err := connection.NewDefaultConnection()
	if err != nil {
		return nil, err
	}
	client := &Client{
		conn:     conn,
		provider: provider,
		authType: authType,
	}

	return client, nil
}
```
and in the `operation` function:
```go
r, err := requests.NewRequest(op, request, authenticator, c.provider)
```

So that we leave all the logic of automatic provider and authenticator selection to the Basic Client, in a simple way? Sorry if that does not work still!!

File: parsec/basicclient.go [hunk 14]
```diff
@@ -12,35 +12,63 @@ import (
...
-	opclient *operations.Client
+	opclient         *operations.Client
+	providerSelector *providerSelector
 }
 
 // InitClient initializes a Parsec client
 func InitClient() (*BasicClient, error) {
-	opclient, err := operations.InitClient()
+	selector := defaultProviderSelector()
+	opclient, err := operations.InitClient(selector)
```
Comment (hug-dev) [inline parsec/basicclient.go:45]: Following my other open comment, `defaultProviderSelector` or similar could instead return a `ProviderID` which it passes to the `InitClient` constructor. It would be something similar as well for the authentication type! I think that avoids going back and forth between the layers.

**Review by hug-dev — APPROVED** [review]
Thank you!! Looks good to me

**Review by ionut-arm — APPROVED** [review]
👌🏻
